[clang-tidy] [dataflow]  Cache reference accessors for `bugprone-unchecked-optional-access`

BaLiKfromUA opened the pull request:
Fixes https://github.com/llvm/llvm-project/issues/126283

Extending https://github.com/llvm/llvm-project/pull/112605 to cache const getters which return references.

Fixes false positives from const reference accessors to object containing optional member

BaLiKfromUA committed:
first implementation and basic tests

BaLiKfromUA committed:
more tests

BaLiKfromUA committed:
format

BaLiKfromUA commented:
TODO: update release notes and `clang-tools-extra/docs/clang-tidy/checks/bugprone/unchecked-optional-access.rst` if needed

BaLiKfromUA commented on clang/lib/Analysis/FlowSensitive/Models/UncheckedOptionalAccessModel.cpp:
```diff
@@ -580,6 +580,26 @@ void handleConstMemberCall(const CallExpr *CE,
+      StorageLocation &Loc =
+          State.Lattice.getOrCreateConstMethodReturnStorageLocation(
+              *RecordLoc, DirectCallee, State.Env, [&](StorageLocation &Loc) {
+                // no-op
```
not sure if anything should be here so decided to make it no-op for now

jvoung replied:
No-op seems okay for this case

BaLiKfromUA commented on clang/lib/Analysis/FlowSensitive/Models/UncheckedOptionalAccessModel.cpp:
```diff
@@ -580,6 +580,26 @@ void handleConstMemberCall(const CallExpr *CE,
+
+    bool isReference =
+        DirectCallee->getReturnType().getTypePtr()->isReferenceType();
+    if (isReference) {
```
maybe this check is redundant

BaLiKfromUA commented on clang/unittests/Analysis/FlowSensitive/UncheckedOptionalAccessModelTest.cpp:
```diff
@@ -3863,6 +3863,192 @@ TEST_P(UncheckedOptionalAccessTest, ConstBoolAccessorWithModInBetween) {
+}
+
+TEST_P(UncheckedOptionalAccessTest,
+       ConstRefToOptionalSavedAsTemporaryVariable) {
```
added this test to check that the existing workaround still works. maybe redundant test

jvoung commented on clang/unittests/Analysis/FlowSensitive/UncheckedOptionalAccessModelTest.cpp:
```diff
@@ -3863,6 +3863,192 @@ TEST_P(UncheckedOptionalAccessTest, ConstBoolAccessorWithModInBetween) {
+  ExpectDiagnosticsFor(R"cc(
+    #include "unchecked_optional_access_test.h"
+
+    class A {
```
Maybe simplify these a little bit (`struct` then don't need the public/private)

jvoung reviewed (commented):
Thanks!

BaLiKfromUA committed:
remove redundant reference check and add one more test

BaLiKfromUA committed:
format

BaLiKfromUA committed:
simplify tests

BaLiKfromUA committed:
update release notes

BaLiKfromUA commented:
Addressed all comments. Thanks for the initial feedback!

Regarding documentation: looks like we need to merge https://github.com/llvm/llvm-project/pull/122290 first and then update docs in my PR

jvoung commented:
> Addressed all comments. Thanks for the initial feedback!
> 
> Regarding documentation: looks like we need to merge #122290 first and then update docs in my PR

Thanks for the reminder about #122290 ! Got back to that and merged. Can you update? It looked like the latest release notes had more entries under "Changes in existing checks" like "bugprone-unsafe-functions" ?

BaLiKfromUA committed:
Merge branch 'main' into issue-126283

BaLiKfromUA committed:
reorder release notes

BaLiKfromUA commented:
Updated release notes. Not sure if I need to update `clang-tidy/checks/bugprone/unchecked-optional-access.rst`. For me it looks like current `Exception: accessor methods` section covers my fix as well.

jvoung commented on clang-tools-extra/docs/ReleaseNotes.rst:
```diff
@@ -114,11 +114,12 @@ Changes in existing checks
   and accessing ``value``. The option `IgnoreSmartPointerDereference` should
   no longer be needed and will be removed.
 
+- Improved :doc:`bugprone-unchecked-optional-access
```
Instead of starting a new bullet, could you append to the above one?

At least from other commits that seems like the precedence: https://github.com/llvm/llvm-project/commit/6504546abcd38159256c3030286b1c02b401c4f8 and https://github.com/llvm/llvm-project/commit/f89fa238faa6a63168997a8a1d03c15b71da8080

===

Also seems like folks wrap to 80-cols

jvoung commented:
> Updated release notes. Not sure if I need to update `clang-tidy/checks/bugprone/unchecked-optional-access.rst`. For me it looks like current `Exception: accessor methods` section covers my fix as well.

Agreed that the "`Exception: accessor methods` section covers" your fix as well, so no need to update. Thank you!

BaLiKfromUA committed:
Update ReleaseNotes.rst

jvoung reviewed (approved):
Thanks!

BaLiKfromUA commented:
Could someone merge it?

Or do we need another reviewer? 

I don't have merge permissions :)
